### ISSUE TITLE
Lazy loading for rust

### ISSUE THREAD

**Issue by AleCandido** [N3PDF/eko#97]
We'd like PineAPPL as well to be able to consume an EKO one `Q2` at a time, since evolving grids requesting a huge number of `Q2` points would require splitting the grid otherwise.

In order to do this, the easiest is to directly provide and maintain a Rust crate that is able to manage the EKO output format.
Indeed, PyO3 would make it possible to run Python code from Rust, but it is more a handle on the interpreter itself, rather than proper bindings in this direction (even though it might be used for this, but it can be painful).

So, the easiest thing it's to make a standalone crate in this repository (unfortunately [`eko`](https://crates.io/crates/eko) is already lost as a name, but we can still use something like *get eko*, i.e. `geko`).
In order to do this, we don't really need many ingredients:
- [ ] a [`tar`](https://crates.io/crates/tar) library, for which the one linked is an obvious candidate
- [ ] a [`yaml`](https://crates.io/crates/yaml-rust) library, that is another easy choice
- [ ] an `npy` library, and here it becomes difficult; the alternatives are
  1. [`npyz`](https://crates.io/crates/npyz) not much maintained, but should be complete enough 
  2. [`ndarray-npy`](https://crates.io/crates/ndarray-npy), slightly more maintained in the recent future, but it does not seem to have a complete feature set
  3. [`npy`](https://crates.io/crates/npy), that is mentioned by the previous, and the number 1. claims to be its fork, but it looks untouched from 2018

**Comment by felixhekhorn** [comment]
as you already said: i. or ii. seem to be likely candidates ...

**Comment by AleCandido** [comment]
Yes, but unfortunately no one of them is very much maintained, so we have to be ready to fix them ourselves (if needed).

We might want to have a look at the code, `.npy` is not a very complicated format, it is just the binary dump of an array + some headers. In any case, the crates are a better starting point than writing everything from scratch, if we're lucky they will completely fulfill our purpose.

**Comment by felixhekhorn** [comment]
maybe we should provide the possibility to do the `xgrid_reshape` on get-time

**Comment by felixhekhorn** [comment]
@alecandido can you please remind me what you meant by "operators headers syncer" above?

**Comment by alecandido** [comment]
It's more than two years ago, but I guess it was:
https://github.com/NNPDF/eko/blob/3bfc89d42c4a501dc7d4ec75aa48f00947e71143/src/eko/io/inventory.py#L186-L203
and friends (e.g. `__iter__`)

**Comment by felixhekhorn** [comment]
ok I see - so for the moment I'm not keeping a copy in memory in Rust: i.e. there is no `Inventory::cache`, just `Inventory::load()` (which, I understood, passes the ownership and thus the responsibility to the calller) - and now I wonder if we really need such a thing ... do we ever read the same 4d operator twice? and if ever can't we rely on the user to keep it alive? I suggest to delay a cache (and any related features) until we have a solid prove we actually need it (which will make the implementation for now much easier)

**Comment by alecandido** [comment]
Well, I guess your main concern is not much the cache implementation, which could be kept in Python, or wherever is needed, but the write operation.
